### DISCUSSION TITLE
9월 27일 인터뷰 피드백

### DISCUSSION THREAD

**Discussion by devHudi** [woowacourse-study/2022-cs-plant#23] [General]
9월 27일 인터뷰이: 후디, 매트

**Comment by hyeonic** [comment]
# 후디 피드백

## 아이스 브레이킹

 * 자기소개
 * 블로그를 하며 가장 크게 얻은 것

## 자바에서 스레드 풀 다뤄보기

- 스레드를 많이 쓸수록 항상 성능이 좋아질까?
- 스레드 풀의 적정 스레드 개수는 몇개라고 생각하는가?
- `ExcutorService`에서 작업 처리 요청을 위해 `excute()`와 `submit()` 메서드를 제공한다. 두 메서드의 차이는 무엇인가?
    - 처리 결과 반환 유무: `execute()`는 `Runnable`만을 처리하며, 작업의 처리 결과를 반환받을 수 없다. 반면, `submit()`은 `Runnable`과 `Callable`두가지를 처리하며 작업의 처리 결과를 `Future`라는 인터페이스로 반환한다.
    - 예외 발생 처리: `execute()`는 쓰레드에서 작업 처리 도중 예외가 발생하면, 쓰레드 풀에서 해당 예외를 제거하고 새로운 쓰레드를 생성한다. 반면, `submit()`은 작업 처리 도중 예외가 발생해도 쓰레드를 제거하지 않고, 다음 작업에 재사용한다. `submit()`이 오버헤드가 더 적으므로 따라서 `submit()`을 사용하는 것이 일반적으로 권장된다.

## 경쟁 상태, 임계 영역의 개념과 동기화를 위한 여러 상호배제 기법

- 경쟁 상태가 발생하는 이유는?
- 뮤텍스와 세마포어의 차이는?
- 자바는 어떠한 방식으로 동기화를 보장하는가?
    - 해당 키워드 말고 보장할 수 있는 또 다른 방법에는 무엇이 있을까?

## 데이터베이스 인덱스와 인덱싱 알고리즘 (hash table, b-tree, b+tree)

- 인덱스 대상 칼럼의 선택 기준은 무엇인가?
- 해시 인덱스 방식의 특징과 부적절한 이유는?
- 인덱싱을 활용한 조회가 불가능한 경우와 그 이유는?

## ****클러스터형 인덱스와 비클러스터형 인덱스****

- 클러스터드 인덱스와 논클러스터드 인덱스의 차이는?
- 클러스터드 인덱스와 논클러스터드 인덱스의 조회 성능에 대한 차이가 있을까?
- 만약 특정 컬럼에 인덱스가 걸려있는 경우 무조건 인덱스를 활용하여 조회를 진행할까?

## **성능 테스트**

- 고가용성이란?
    - 서버와 네트워크, 프로그램 등의 정보 시스템이 상당히 오랜 시간 동안 지속적으로 정상 운영이 가능한 형질을 이야기 한다.
- TPS가 높다는 것은 어떤 것을 의미하는가?

**Comment by devHudi** [comment]
- 뮤텍스, 세마포어 차이점과 메커니즘 더 깊게 공부해야할 것 같다.
- 클러스터형 인덱스와 비클러스터형 인덱스의 동작 메커니즘을 더 자세히 공부해야할 것 같다.
- 쓰레드 컨텍스트 스위칭에 대해 공부해야할 것 같다.
- 물리 코어와 쓰레드에 대해 공부해야할 것 같다.

**Comment by devHudi** [comment]
# 매트

## **질문**

### **트랜잭션과 격리 레벨**

> 이슈 링크 : https://github.com/woowacourse-study/2022-cs-plant/issues/13
> 
- 트랜잭션을 왜 사용하나요? 사용하지 않으면 어떤 일이 발생하나요?
- 트랜잭션 4가지 특징을 간단히 설명해주세요.
    - ACID에서 ‘일관성’ 이라는 특징이 있는데, 데이터베이스가 일관적이다라는게 무슨 뜻일까요?
        - 제약 조건이 무엇인가요?
- 커밋되지 않은 읽기(READ UNCOMMITED)는 굉장히 위험해보이는데 사용할 필요가 있나요?
    - 사용할만한 사례가 뭔가 있을까요?
- 언두 영역에 대해서 소개해주세요.
- 커밋된 읽기(READ COMMITED)에서 반복 불가능한 읽기(NON REPEATABLE READ)가 발생하는 이유가 무엇인가요?
- 프로젝트에서 트랜잭션 격리 레벨을 도입한다면 어디에 도입하실 것인가요?

### 클러스터링 인덱스와 논 클러스터링 인덱스

> 이슈 링크 : https://github.com/woowacourse-study/2022-cs-plant/issues/14
> 
- 데이터베이스 클러스터링 을 왜 하는걸까요? 그냥 데이터를 전부 메모리에 올리면 안되나요?
- 클러스터링 인덱스와 논 클러스터링 인덱스가 혼합되어 사용되었을 경우 데이터를 탐색하는 과정에 대해 설명해주세요.
    
    > 논 클러스터링 인덱스를 통해 먼저 데이터를 찾는다. 이때, 논 클러스터링 인덱스의 리프 노드에는 데이터 페이지 번호와 오프셋 대신 클러스터링 인덱스의 키를 가지고 있다. 이 클러스터링 인덱스의 키를 사용하여 클러스터링 인덱스로 데이터에 접근한다.
    > 

### **인덱스 스캔**

> 이슈 링크 : https://github.com/woowacourse-study/2022-cs-plant/issues/15
> 
- 인덱스 레인지 스캔과 풀 스캔이 가능한 이유는 데이터베이스 인덱스가 어떤 자료구조를 사용하기 때문인가요?
    
    > (이건 개인적인 생각이 절반임) B+Tree. B-Tree는 리프 노드가 링크드 리스트 형태로 이어지지 않으므로 수평적인 탐색이 불가능하다. 정확히는 B+Tree가 정답. (Real MySQL에서 B-Tree라고 설명되어 있으나, 실제로 그림에서 표현하고 있는 트리의 리프 노드는 서로 이어져있으므로 B+Tree라고 설명해야 더 정확하다… 고 생각하지만 개인적인 의견임으로 적당히 걸러 들어야할 듯) (참고: [https://hwannny.tistory.com/101](https://hwannny.tistory.com/101))
    > 
- 인덱스 풀 스캔은 언제 발생하나요?
    
    > 인덱스가 A, B, C 순서로 생성되어 있는데, WHERE 절에 B나 C칼럼이 사용되는 등 쿼리 조건절에 사용된 컬럼이 인덱스의 첫번째 컬럼이 아닌 경우.
    > 
- 커버링 인덱스가 무엇인가요?
    
    > [https://tecoble.techcourse.co.kr/post/2021-10-12-covering-index/](https://tecoble.techcourse.co.kr/post/2021-10-12-covering-index/)
    > 

### **Array**

> 이슈 링크 : https://github.com/woowacourse-study/2022-cs-plant/issues/19
> 
- ArrayList와 LinkedList의 차이를 요소 조회와 삽입/삭제의 관점에서 설명해주세요.
    - 삽입/삭제 시간 복잡도 비교
- 자바에서 제공하는 ArrayList의 요소들은 실제로 물리적으로 연속되어 저장되나요?
- 자바의 ArrayList는 가변적인 크기를 제공합니다. 그런데, 어떻게 항상 모든 요소들이 연속되어 저장되나요?
    
    > ArrayList의 내부 배열이 부족해지면, `grow()` 메소드를 호출하는데 이때 `Arrays.copyOf()` 를 사용해서 아예 새로운 배열을 만든다.
    참고 : [https://jaehoney.tistory.com/184](https://jaehoney.tistory.com/184)
    > 

### **HashMap**

> 이슈 링크 : https://github.com/woowacourse-study/2022-cs-plant/issues/20
> 
- 해시 충돌과 원인에 대해서 설명해주세요.
- 완전 해시함수에 대해서 설명해주세요.
    
    > 모든 키가 다른 해시 값을 가지게 되어 해시 충돌이 발생하지 않는 해시 함수 ([https://zetawiki.com/wiki/완전_해시함수](https://zetawiki.com/wiki/%EC%99%84%EC%A0%84_%ED%95%B4%EC%8B%9C%ED%95%A8%EC%88%98))
    > 
- HashTable이 Thread-Safe 하다면 특히 멀티쓰레드를 사용하는 웹 어플리케이션에서는 HashMap 보다 더 좋을 것 같은데, 왜 사용이 권장되지 않는건가요?
    
    > 멀티 쓰레드 환경에서는 동기화가 보장되며, 성능도 더 우수한 ConcurrentHashMap을 사용하면 된다. ConcurrentHashMap 같은 경우 조작하게 되는 특정 Entry에 대해서만 락을 걸기 때문에 HashTable 대비 성능이 더 우수하다.
    > 
- 왜 String과 POJO는 완벽한 해시 함수를 만드는게 불가능한가요?
    
    > 나도 잘 모르겠음
    > 
- Open Addressing 방식에서 데이터를 추가하려는데 비어있는 버킷이 없다면 어떻게 되나요?
    
    > 버킷을 확장한다. 다만, 이 작업은 굉장히 오래걸린다.
    > 
- 적은 데이터를 저장하는 해시 테이블을 직접 구현한다면, Open Addressing 방식으로 구현하실건가요? 아니면 Chaining 방식을 사용하실 건가요?
    
![Untitled](https://user-images.githubusercontent.com/11745691/192518763-8b693fa7-0aac-4169-822a-d2aa762c4237.png)

    
    > 저장되는 데이터가 적다면 해시 충돌의 확률이 낮아진다. 따라서 Open Addressing 방식을 사용하면 List 혹은 Tree를 가리키는 포인터에 대한 메모리를 아낄 수 있다. 또한 버킷 내부의 또 다른 자료구조를 탐색하는 작업도 필요 없어진다.
    > 
- 자바의 HashMap은 왜 내부적으로 LinkedList와 Tree 두가지 자료구조를 사용하나요?

## **피드백**

- 여유로워 보였다.
- 경직되어 보였다.
- 주어진 정보들에서 잘 찾아가는 느낌.

**Comment by gudonghee2000** [comment]
## 후디 
----
### 피드백
- 대화하듯이 자연스럽고 편안한 분위기였다.
- 제스처를 통해서 이야기하는 모습이 좋았다.
- 마지막 질문에서 일리가 있는 부분들에 대해서 자신의 생각을 이야기 하는부분이 인상깊었다. 
- 기억이 잘나지 않거나 헷갈리는 부분에 대해서 시간을 달라고 하는 모습이 좋았다.

**Comment by gudonghee2000** [comment]
## 매트
----
### 피드백
- 계속적으로 연결되는 질문들에 대해서 막히지 않고 대답하는 모습이 좋았다.
- 대화하듯이 편안한 분위기였다.
- 제스처를 포함하여 질문에 답변을 해준다면 더 좋을것 같다.
- 생각하지 못한 질문에 대해서 고민하고 정답을 찾아가는 모습이 인상깊었다.
